_posts/2023-09-06-load-balancer.md

```diff
@@ -0,0 +1,906 @@
+backends=[
+ "http://localhost:8081",
+ "http://localhost:8082",
+ "http://localhost:8083"
+], // backend server URL-s
+health-check-interval="1" // check backend health per every new second
+```
+
+Now we can create the `Config` model for capturing the aforementioned idea:
+
+```scala
+package com.ghurtchu.loadbalancer.domain
+
+import com.ghurtchu.loadbalancer.domain.Config.HealthCheckInterval
+
+import scala.util.Try
+
+final case class Config(
+  port: String,
+  host: String,
+  backends: Urls,
+  healthCheckInterval: HealthCheckInterval
+) {
+
+  def hostOr(fallback: String): String =
+    if (host.isEmpty) fallback
+    else host
+
+  def portOr(fallback: Int): Int =
+    Try(port.toInt).toOption
+      .getOrElse(fallback)
+}
+
+object Config {
+
+  type InvalidConfig = InvalidConfig.type
+
+  final case object InvalidConfig extends Throwable {
+    override def getMessage: String =
+      "Invalid port or host, please fix Config"
+  }
+
+  final case class HealthCheckInterval(value: Long) extends AnyVal
+}
+```
+
+And we can also write a few tests for `Config` to verify its public API:
+```scala
+package com.ghurtchu.loadbalancer.domain
+
+import munit.FunSuite
+import Config.HealthCheckInterval
+
+class ConfigTest extends FunSuite {
+
+  val config = Config(
+    port = "8081",
+    host = "localhost",
+    backends = Urls.empty,
+    healthCheckInterval = HealthCheckInterval(1),
+  )
+
+  test("hostOr") {
+    val obtained = config.hostOr("0.0.0.0")
+    val expected = "localhost"
+    assertEquals(obtained, expected)
+
+    val configWithEmptyHost = config.copy(host = "")
+    val obtainedDefault     = configWithEmptyHost.hostOr("0.0.0.0")
+    val expectedDefault     = "0.0.0.0"
+    assertEquals(obtainedDefault, expectedDefault)
+  }
+
+  test("portOr") {
+    val obtained = config.portOr(8080)
+    val expected = 8081
+    assertEquals(obtained, expected)
+
+    val configWithEmptyPort = config.copy(port = "invalid port")
+    val obtainedDefault     = configWithEmptyPort.portOr(8080)
+    val expectedDefault     = 8080
+    assertEquals(obtainedDefault, expectedDefault)
+  }
+}
+```
+
+Splendid!
+
+Ok, what about the fact that our load balancer should incorporate concurrency and thread-safety?
+
+It is apparent that the load balancer will have to handle concurrent requests from different clients which can be issued even the same time! So we need to make sure that the state updates for 
+`Urls` is atomic.
+
+For that we can make use of `cats.effect.Ref`, however we need to distinguish between two completely separate problems:
+- forwarding requests to the backends and returning the response to the clients (main feature)
+- checking the health of the backends (secondary feature)
+
+So, clearly we see that we need to have two separate instances which wrap `cats.effect.Ref`. Let's name them as `Backends` and `HealthChecks`:
+
+```scala
+package com.ghurtchu.loadbalancer.domain
+
+import cats.effect.{IO, Ref}
+
+trait UrlsRef {
```

Comment:
just make this a sealed trait and place the subtypes top-level, otherwise imports are weird

Reply:
You mean this way?

```scala
sealed trait UrlsRef {
  def urls: Ref[IO, Urls]
}


final case class Backends(urls: Ref[IO, Urls])     extends UrlsRef
final case class HealthChecks(urls: Ref[IO, Urls]) extends UrlsRef
```

Reply:
Now it will be Scala 3 `enum`